operators/pkg/controller/elasticsearch/elasticsearch_controller.go

```diff
@@ -283,3 +297,78 @@ func (r *ReconcileElasticsearch) finalizersFor(
 		http.DynamicWatchesFinalizer(r.dynamicWatches, es.Name, esname.ESNamer),
 	}
 }
+
+// reconcileCompatibility determines if this controller is compatible with a given resource by examining the controller version annotation
+// controller versions 0.9.0+ cannot reconcile resources created with earlier controllers, so this lets our controller skip those resources until they can be manually recreated
+// if an object does not have an annotation, it will determine if it is a new object or if it has been previously reconciled by an older controller version, as this annotation
+// was not applied by earlier controller versions. it will update the object's annotations indicating it is incompatible if so
+func (r *ReconcileElasticsearch) reconcileCompatibility(es *elasticsearchv1alpha1.Elasticsearch) (bool, error) {
+	annExists := es.Annotations != nil && es.Annotations[annotation.ControllerVersionAnnotation] != ""
+
+	// if the annotation does not exist, it might indicate it was reconciled by an older controller version that did not add the version annotation,
+	// in which case it is incompatible with the current controller, or it is a brand new resource that has not been reconciled by any controller yet
+	if !annExists {
+		exist, err := r.checkExistingResources(es)
+		if err != nil {
+			return false, err
+		}
+		if exist {
+			log.Info("Resource was previously reconciled by incompatible controller version and missing annotation, adding annotation", "controller_version", r.OperatorInfo.BuildInfo.Version, "namespace", es.Namespace, "es_name", es.Name)
+			err = annotation.UpdateControllerVersion(r.Client, es, "0.8.0-UNKNOWN")
```

Comment:
Would it make sense to also set the `common.k8s.elastic.co/pause: true` annotation as well so that the association controllers stop working on these old resources. Alternatively implement version check on the annotation controllers as well

Reply:
I went with adding the version check to the kibana association controller -- I don't think the APM association controller needs it since it will only act on APM servers, which don't exist < 0.9.0 right?

Reply:
We did include APM server in 0.8 it was just not announced because the implementation was incomplete